core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala

```diff
@@ -16,233 +16,1521 @@
+    val nonExistingBrokerIDs = proposedParts.toMap.values.flatten.filterNot(existingBrokerIDs.contains).toSet
+    if (nonExistingBrokerIDs.nonEmpty)
+      throw new AdminCommandFailedException("The proposed assignment contains non-existent brokerIDs: " + nonExistingBrokerIDs.mkString(","))
+  }
+
+  /**
+   * Execute the given partition reassignments.
+   *
+   * @param adminClient       The admin client object to use.
+   * @param reassignments     A map from topic names to target replica assignments.
+   * @return                  A map from partition objects to error strings.
+   */
+  def alterPartitionReassignments(adminClient: Admin,
+                                  reassignments: Map[TopicPartition, Seq[Int]])
+                                  : Map[TopicPartition, Throwable] = {
+    val results: Map[TopicPartition, KafkaFuture[Void]] =
+      adminClient.alterPartitionReassignments(reassignments.map {
+        case (part, replicas) => {
+          (part, Optional.of(new NewPartitionReassignment(replicas.map(Integer.valueOf(_)).asJava)))
+        }
+      }.asJava).values().asScala
+    results.flatMap {
+      case (part, future) => {
+        try {
+          future.get()
+          None
+        } catch {
+          case t: ExecutionException => Some(part, t.getCause())
+        }
+      }
+    }
+  }
+
+  /**
+   * Cancel the given partition reassignments.
+   *
+   * @param adminClient       The admin client object to use.
+   * @param reassignments     The partition reassignments to cancel.
+   * @return                  A map from partition objects to error strings.
+   */
+  def cancelPartitionReassignments(adminClient: Admin,
+                                  reassignments: Set[TopicPartition])
+  : Map[TopicPartition, Throwable] = {
+    val results: Map[TopicPartition, KafkaFuture[Void]] =
+      adminClient.alterPartitionReassignments(reassignments.map {
+          (_, (None: Option[NewPartitionReassignment]).asJava)
+        }.toMap.asJava).values().asScala
+    results.flatMap {
+      case (part, future) => {
+        try {
+          future.get()
+          None
+        } catch {
+          case t: ExecutionException => Some(part, t.getCause())
+        }
+      }
+    }
+  }
+
+  /**
+   * Calculate the global map of all partitions that are moving.
+   *
+   * @param currentReassignments    The currently active reassignments.
+   * @param proposedReassignments   The proposed reassignments (destinations replicas only).
+   * @param currentParts            The current location of the partitions that we are
+   *                                proposing to move.
+   * @return                        A map from topic name to partition map.
+   *                                The partition map is keyed on partition index and contains
+   *                                the movements for that partition.
+   */
+  def calculateMoveMap(currentReassignments: Map[TopicPa
```

Comment:
Do we allow cancellation only of partition reassignments whose `addingReplicas` isn't empty or whose `removingReplicas` isn't empty? If so, the latter check needs a `!`

What about the alternative - allow cancellation of whatever `listPartitionReassignments()` returns? The API should return reassignments that are in progress only anyway

Reply:
Hmm, github is showing this comment in a weird place-- on the `calculateMoveMap` function rather than in the `cancelAssignment` function as you probably intended.

Good find on the missing negation.  That was definitely not intended.  Fixed.

If users want to cancel arbitrary partition moves, they can create a JSON file and use `--cancel` on it, right?